### PR TITLE
[Sampler] BiasedChoice sampler

### PR DESCRIPTION
**Pull request by soodoshll**
## Description

Fundamental functions for biased sampling. 

- Adapt tree sampler to enable removing an element from a group within which elements have the same probability.
- Implement BiasedChoice based on that.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

**Comment by yzh119** [comment]
@soodoshll is this PR ready for review?

**Comment by soodoshll** [comment]
> @soodoshll is this PR ready for review?

Yeah

**Review by jermainewang — CHANGES_REQUESTED** [review]

File: include/dgl/random.h [hunk 1]
```diff
@@ -178,6 +178,42 @@ class RandomEngine {
...
+   * \brief Pick random integers with different probability for different segments.
+   *
+   * If replace is false, num must not be larger than population.
+   *
+   * \tparam IdxType Return integer type
+   * \param num Number of integers to choose
+   * \param split Split positions of different segments
+   * \param bias Weights of different segments
+   * \param out The output buffer to write selected indices.
+   * \param replace If true, choose with replacement.
```
[the remark this answers is not in the document]
Comment (soodoshll) [reply include/dgl/random.h:206]: The `split` arrays of each node are compactly stored in a N*(T+1) array, where N is the number of nodes and T is the number of tags. And the input`split` is obtained by pointer displacement.

We need the `shape` attribute of `bias` to determine how many segments there are.

File: src/random/cpu/sample_utils.h [hunk 2]
```diff
@@ -279,7 +279,18 @@ class TreeSampler: public BaseSampler<Idx> {
...
+   * The parameter decrease is an array of which the length is the number of categories.
+   * Every time an element in the category x is picked, the weight of this category is subtracted
+   * by decrease[x]. It is used to support the case where a category might contains multiple
+   * candidates and decrease[x] is the weight of one candidate of the category x.
+   *
+   * When decrease == nullptr, it means there is only one candidate in each category and will
+   * directly set the weight of the chosen category as 0.
+   *
+   */
+  Idx DrawAndUpdate(const DType *decrease = nullptr) {
```
Comment (jermainewang) [inline src/random/cpu/sample_utils.h:293]: why not just call it `Draw`? this interface should be compatible with the old one right?

**Review by jermainewang — APPROVED** [review]

File: src/random/cpu/sample_utils.h (hunk 2 above)
Comment (soodoshll) [reply src/random/cpu/sample_utils.h:293]: We need to also modify the interface of `BaseSampler` to make the interface consistent with the super class. But `BaseSampler` does not have template parameter `DType`.

**Pull request merged**